corehq/apps/app_manager/models.py

```diff
@@ -4911,6 +4911,9 @@ def _hash(val):
                     my_hash = _hash(self.fetch_xform(form))
                     if previous_hash != my_hash:
                         form.version = None
+                        # fetch_xform calls render_xform which is memoized
```

Comment:
Can this function use `render_xform` directly instead of `fetch_xform`? Using fetch_xform calls validate_form, which is slow and shouldn't be necessary at this point - the app has already been validated.

Reply:
> the app has already been validated

Has it? At least in the context of this request, my understanding is this is the first time we are interacting with the form via `create_all_files`. Though I admittedly have a limited understanding of app manager code/flows in general.

Reply:
Ah nevermind. I see it gets called in `get_direct_ccz` pretty much immediately.

Reply:
yeah, validation should have been completed before any file-creating happens. Historically, we've sometimes ended up with redundant validation code, which can then be low-hanging fruit to improve performance.

Reply:
Coming back to this, so I modified this a bit but not in the way your comment initially calls out. Because the call to `set_form_versions` is the first thing that happens in `create_all_files`, it appears to also be responsible for validation. So I've separated out the validation and render calls just to be more explicit (tempted to do the same for the last call to `fetch_xform` and remove that method entirely to avoid redundant validation in the future). However in the later call to `fetch_xform` that happens in `_get_form_files`, I've only called `render_xform` like you mentioned above with no more validation since it is has been done. How does that sound to you?

Reply:
I think the new commit is good...but there's literally only one call to `fetch_xform` now, can you just kill the function?

>  Because the call to set_form_versions is the first thing that happens in create_all_files, it appears to also be responsible for validation.

This is distressing, and I'd much rather see validation done explicitly as the first step in `create_all_files`, but up to you if that's out of scope.

Reply:
> I think the new commit is good...but there's literally only one call to fetch_xform now, can you just kill the function?

Yeah I agree it makes sense to remove. See ed9373a48ee94e851c2c3ec8550c81fc56be0b6e

Reply:
Sounds fine. Regarding `render_xform`, you can remove the entire definition now, can't you?

Reply:
🤦🏻 yeah forgot to do that. Thanks.

Reply:
Created https://dimagi-dev.atlassian.net/browse/SAAS-13392